src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java

```diff
@@ -853,109 +853,96 @@ public void flush(Flush flush) throws EngineException {
             throw new FlushNotAllowedEngineException(shardId, "already flushing...");
         }
 
-        flushLock.lock();
-        try {
+        final InternalLock lockNeeded;
+        switch (flush.type()) {
+            case NEW_WRITER:
+                lockNeeded = writeLock;
+                break;
+            case COMMIT:
+            case COMMIT_TRANSLOG:
+                lockNeeded = readLock;
+                break;
+            default:
+                throw new ElasticsearchIllegalStateException("flush type [" + flush.type() + "] not supported");
+        }
+
+        try (InternalLock _ = lockNeeded.acquire(); InternalLock flock = flushLock.acquire()) {
```

Comment:
took me a while to get what you are doing here but I like it. Can you put a big comment here that we are doing this in a certain order to prevent deadlocks?